torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <c10/util/Exception.h>
+#include <torch/data/datasets/base.h>
+#include <torch/data/example.h>
+#include <torch/data/samplers.h>
+
+namespace torch {
+namespace data {
+namespace datasets {
+/// A stateful dataset that support hierarchical sampling and prefetching of
+/// entre chunks. dataset that supports loading an entire chunk of data.
+///
+/// A chunk could be an entire file, such as an audio data file or an image,
+/// or part of a file in the case of a large text file split based on seek
+/// positions.
+///
+/// Unlike regular dataset, chunk dataset require two samplers to operate and
+/// keeps an internal state. `ChunkSampler` selects, which chunk to load next,
+/// while the `ExampleSampler` determins the order of Examples that are returned
+/// in each `get_batch` call. The hierarchical sampling approach used here is
+/// inspired by this paper http://martin.zinkevich.org/publications/nips2010.pdf
+template <
+    typename Self,
+    typename Batch = std::vector<Example<>>,
+    typename BatchRequest = ArrayRef<size_t>,
+    typename ChunkSampler = samplers::RandomSampler,
+    typename ExampleSampler = samplers::RandomSampler>
+class ChunkDataSet : public BatchDataset<Self, Batch, BatchRequest> {
+ public:
+  ~ChunkDataSet() {
```

Comment:
Let's make this a virtual destructor for future inheritance.

Reply:
done.